src/thunderbots/software/ai/passing/evaluation.cpp

```diff
@@ -2,14 +2,297 @@
+
+    // Figure out the range of angles for which we have an open shot to the goal after
+    // receiving the pass
+    Angle open_angle_to_goal =
+        angleSweepCircles(pass.receiverPoint(), field.enemyGoalpostNeg(),
+                          field.enemyGoalpostPos(), obstacles, ROBOT_MAX_RADIUS_METERS)
+            .second;
+
+    // Create the shoot score by creating a sigmoid that goes to a large value as
+    // we get to the ideal shoot angle.
+    double shot_openness_score =
+        sigmoid(open_angle_to_goal.toDegrees(), 0.5 * ideal_shoot_angle_degrees,
+                ideal_shoot_angle_degrees);
+
+    // Prefer angles where the robot does not have to turn much after receiving the
+    // pass to take the shot
+    Angle pass_orientation = (pass.passerPoint() - pass.receiverPoint()).orientation();
+    Angle post0_diff       = pass_orientation.minDiff(
+        (field.enemyGoalpostNeg() - pass.receiverPoint()).orientation());
+    Angle post1_diff = pass_orientation.minDiff(
+        (field.enemyGoalpostPos() - pass.receiverPoint()).orientation());
+    Angle min_rotation_to_shot_after_pass   = std::min(post0_diff, post1_diff);
+    double required_rotation_for_shot_score = sigmoid(
+        min_rotation_to_shot_after_pass.toDegrees(),
+        0.5 * ideal_min_rotation_to_shoot_degrees, ideal_min_rotation_to_shoot_degrees);
+
+    return shot_openness_score * required_rotation_for_shot_score;
+}
+
+double AI::Passing::ratePassEnemyRisk(const Team& enemy_team, const Pass& pass)
+{
+    double enemy_proximity_importance =
+        Util::DynamicParameters::AI::Passing::enemy_proximity_importance.value();
+
+    // Calculate a risk score based on the distance of the enemy robots from the receive
+    // point, based on an exponential function of the distance of each robot from the
+    // receiver point
+    auto enemy_robots                    = enemy_team.getAllRobots();
+    double enemy_receiver_proximity_risk = 1;
+    for (const Robot& enemy : enemy_team.getAllRobots())
+    {
+        double dist = (pass.receiverPoint() - enemy.position()).len();
+        enemy_receiver_proximity_risk *=
+            enemy_proximity_importance * std::exp(-dist * dist);
+    }
+    if (enemy_robots.empty())
+    {
+        enemy_receiver_proximity_risk = 0;
+    }
+
+    double intercept_risk = calculateInterceptRisk(enemy_team, pass);
+
+    // We want to rate a pass more highly if it is lower risk, so subtract from 1
+    return 1 - std::max(intercept_risk, enemy_receiver_proximity_risk);
+}
+
+double AI::Passing::calculateInterceptRisk(const Team& enemy_team, const Pass& pass)
+{
+    // Return the highest risk for all the enemy robots, if there are any
+    auto enemy_robots = enemy_team.getAllRobots();
+    if (enemy_robots.empty())
+    {
+        return 0;
+    }
+    std::vector<double> enemy_intercept_risks;
+    std::transform(enemy_robots.begin(), enemy_robots.end(),
+                   std::back_inserter(enemy_intercept_risks),
```

Comment:
nit: back_inserter could cause reallocs, you could call the constructor of std::vector with `enemy_robots.size()` instead.

Reply:
Good point, changed!